src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1046,300 @@ begin
+linear_map.ker_eq_bot.2 (to_lin_injective hB)
+
+/-- Let `B` be a symmetric, nondegenerate bilinear form on a nontrivial module `M` over the ring
+  `R` with invertible `2`. Then, there exists some `x : M` such that `B x x ≠ 0`. -/
+lemma exists_bilin_form_self_neq_zero [htwo : invertible (2 : R)] [nontrivial M]
+  {B : bilin_form R M} (hB₁ : B.nondegenerate) (hB₂ : sym_bilin_form.is_sym B) :
+  ∃ x, ¬ B.is_ortho x x :=
+begin
+  by_contra, push_neg at h,
+  have : ∀ u v, 2 * B u v = 0,
+  { intros u v,
+    rw [show 2 * B u v = B u u + B v u + B u v + B v v,
+          by rw [is_ortho_def.1 (h u), is_ortho_def.1 (h v), hB₂ v u, two_mul, zero_add]; ring,
+        show B u u + B v u + B u v + B v v = B (u + v) (u + v),
+          by simp [← add_assoc], is_ortho_def.1 (h _)] },
+  have hcon : ∀ u v, B u v = 0,
+  { intros u v,
+    rw [show 0 = htwo.inv_of * (2 * B u v), by rw [this, mul_zero]], simp [← mul_assoc] },
+  exact let ⟨v, hv⟩ := exists_ne (0 : M) in hv $ hB₁ v (hcon v),
+end
+
+variables {V : Type u} {K : Type v}
+variables [field K] [add_comm_group V] [vector_space K V]
+
+/-- A set of orthogonal vectors `v` with respect to some bilinear form `B` is  linearly independent
+  if for all `i`, `B (v i) (v i) ≠ 0`. -/
+lemma linear_independent_of_is_Ortho
+  {n : Type w} {B : bilin_form K V} {v : n → V}
+  (hv₁ : B.is_Ortho v) (hv₂ : ∀ i, ¬ B.is_ortho (v i) (v i)) : linear_independent K v :=
+begin
+  classical,
+  rw linear_independent_iff',
+  intros s w hs i hi,
+  have : B (s.sum $ λ (i : n), w i • v i) (v i) = 0,
+  { rw [hs, zero_left] },
+  have hsum : s.sum (λ (j : n), w j * B (v j) (v i)) =
+    s.sum (λ (j : n), if i = j then w j * B (v j) (v i) else 0),
+  { refine finset.sum_congr rfl (λ j hj, _),
+    by_cases (i = j),
+    { rw [if_pos h] },
+    { rw [if_neg h, is_Ortho_def.1 hv₁ _ _ h, mul_zero] } },
+  simp_rw [map_sum_left, smul_left, hsum, finset.sum_ite_eq] at this,
+  rw [if_pos, mul_eq_zero] at this,
+  cases this,
+  { assumption },
+  { exact false.elim (hv₂ i $ this) },
+  { assumption }
+end
+
+-- ↓ This lemma only applies in fields as we require `a * b = 0 → a = 0 ∨ b = 0`
+lemma span_inf_ortho_eq_bot {B : bilin_form K V} {x : V} (hx : ¬ B.is_ortho x x) :
+  (K ∙ x) ⊓ B.orthogonal (K ∙ x) = ⊥ :=
+begin
+  rw ← finset.coe_singleton,
+  refine eq_bot_iff.2 (λ y h, _),
+  rcases mem_span_finset.1 h.1 with ⟨μ, rfl⟩,
+  have := h.2 x _,
+  { rw finset.sum_singleton at this ⊢,
+    suffices hμzero : μ x = 0,
+    { rw [hμzero, zero_smul, submodule.mem_bot] },
+    change B (μ x • x) x = 0 at this, rw [smul_left] at this,
+    exact or.elim (zero_eq_mul.mp this.symm) id (λ hfalse, false.elim $ hx hfalse) },
+  { rw submodule.mem_span; exact λ _ hp, hp $ finset.mem_singleton_self _ }
+end
+
+-- ↓ This lemma only applies in fields since we use the inverse
+lemma span_sup_ortho_eq_top {B : bilin_form K V}
+  (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :
+  (K ∙ x) ⊔ B.orthogonal (K ∙ x) = ⊤ :=
```

Comment:
Do we need the hypothesis of symmetry here?  Can you frame a general lemma:  let `V` be a finite-dimensional vector space, let `L : V →ₗ[K] K` be a linear map, and let `x : V` be such that `L x` is nonzero; then `(K ∙ x) ⊔ L.ker = ⊤`.

Reply:
Right, that is certainly doable. However, I did not assume finite-dimensional in this lemma, do you think this lemma for linear maps is worth while having as a separate thing?

Reply:
Ha ... maybe "finite-dimensional" is also unnecessary in the lemma I stated? :)

Reply:
The proof I have in mind right now uses the first isomorphism theorem and requires finite-dimensional. I will try to come up with something else

Reply:
I don't think the statement is true for infinite-dimensional vector spaces. If we consider the linear map $L : C^1(\R) \to \R$ where $L$ is differentiation evaluated at 0. Then $L(2x) \neq 0$ but clearly $span(2x) + ker L \neq C^1(\R)$.

Reply:
Sorry, I don't quite follow -- given `f` in `C^1(R)`, we can write `f = (1/2) f'(0) ∙ (2x) + [f - (1/2) f'(0) ∙ (2x)]`, and isn't the latter in the kernel of `L`?

Reply:
Sorry, my mental image way clearly wrong in my example. I believe I have now come up with a math proof. Let `y : V` be arbitrary and choose `l = L y / L x`. Then `0 = L y - L (l * x) = L(y - l * x)` so `y + l * x \in ker L`. Hence, `y = l * x + y - l * x`. I will write this up in lean as soon as I'm back on a computer :)

Reply:
I have now done this with `span_singleton_sup_ker_eq_top`